lighthouse-core/report/html/renderer/element-screenshot-renderer.js

```diff
@@ -0,0 +1,259 @@
+      height: dom.document().documentElement.clientHeight - reportEl.getBoundingClientRect().top,
+    };
+    for (const el of dom.document().querySelectorAll('.lh-element-screenshot')) {
+      el.addEventListener('click', () => {
+        const overlay = dom.createElement('div');
+        overlay.classList.add('lh-element-screenshot__overlay');
+        const elementRectInScreenshotCoords = {
+          width: Number(el.getAttribute('rectWidth')),
+          height: Number(el.getAttribute('rectHeight')),
+          left: Number(el.getAttribute('rectLeft')),
+          right: Number(el.getAttribute('rectLeft')) + Number(el.getAttribute('rectWidth')),
+          top: Number(el.getAttribute('rectTop')),
+          bottom: Number(el.getAttribute('rectTop')) + Number(el.getAttribute('rectHeight')),
+        };
+        overlay.appendChild(ElementScreenshotRenderer.render(
+          dom,
+          templateContext,
+          fullPageScreenshot,
+          elementRectInScreenshotCoords,
+          renderContainerSizeInDisplayCoords
+        ));
+        overlay.addEventListener('click', () => {
+          overlay.remove();
+        });
+
+        reportEl.appendChild(overlay);
+      });
+    }
+  }
+
+  /**
+   * Given the size of the element in the screenshot and the total available size of our preview container,
+   * compute the factor by which we need to zoom out to view the entire element with context.
+   * @param {LH.Artifacts.Rect} elementRectInScreenshotCoords
+   * @param {Size} renderContainerSizeInDisplayCoords
+   * @return {number}
+   */
+  static _computeZoomFactor(elementRectInScreenshotCoords, renderContainerSizeInDisplayCoords) {
+    const targetClipToViewportRatio = 0.75;
+    const zoomRatioXY = {
+      x: renderContainerSizeInDisplayCoords.width / elementRectInScreenshotCoords.width,
+      y: renderContainerSizeInDisplayCoords.height / elementRectInScreenshotCoords.height,
+    };
+    const zoomFactor = targetClipToViewportRatio * Math.min(zoomRatioXY.x, zoomRatioXY.y);
+    return Math.min(1, zoomFactor);
+  }
+
+  /**
+   * Renders an element with surrounding context from the full page screenshot.
+   * Used to render both the thumbnail preview in details tables and the full-page screenshot in the lightbox.
+   * @param {DOM} dom
+   * @param {ParentNode} templateContext
+   * @param {LH.Artifacts.FullPageScreenshot} fullPageScreenshot
+   * @param {LH.Artifacts.Rect} elementRectInScreenshotCoords Region of screenshot to highlight.
+   * @param {Size} renderContainerSizeInDisplayCoords
+   * @return {Element}
+   */
+  static render(dom, templateContext, fullPageScreenshot, elementRectInScreenshotCoords,
+      renderContainerSizeInDisplayCoords) {
+    //
+    const tmpl = dom.cloneTemplate('#tmpl-lh-element-screenshot', templateContext);
+    const containerEl = dom.find('.lh-element-screenshot', tmpl);
+
+    containerEl.setAttribute('rectWidth', elementRectInScreenshotCoords.width.toString());
```

Comment:
data attributes may be a bit cleaner? Or just properties?

Reply:
we use data attributes for all the `node` deets. i'd prefer using em here too.

small plus is we can use `elem.dataset`